### PR TITLE
Rewrite and refactor format_args!() builtin macro.

### PR DESCRIPTION
**Pull request by m-ou-se**
This is a near complete rewrite of `compiler/rustc_builtin_macros/src/format.rs`.

This gets rid of the massive unmaintanable [`Context` struct](https://github.com/rust-lang/rust/blob/76531befc4b0352247ada67bd225e8cf71ee5686/compiler/rustc_builtin_macros/src/format.rs#L176-L263), and splits the macro expansion into three parts:

1. First, `parse_args` will parse the `(literal, arg, arg, name=arg, name=arg)` syntax, but doesn't parse the template (the literal) itself.
2. Second, `make_format_args` will parse the template, the format options, resolve argument references, produce diagnostics, and turn the whole thing into a `FormatArgs` structure.
3. Finally, `expand_parsed_format_args` will turn that `FormatArgs` structure into the expression that the macro expands to.

In other words, the `format_args` builtin macro used to be a hard-to-maintain 'single pass compiler', which I've split into a three phase compiler with a parser/tokenizer (step 1), semantic analysis (step 2), and backend (step 3). (It's compilers all the way down. ^^)

This can serve as a great starting point for https://github.com/rust-lang/rust/issues/99012, which will only need to change the implementation of 3, while leaving step 1 and 2 unchanged.

It also makes https://github.com/rust-lang/compiler-team/issues/541 easier, which could then upgrade the new `FormatArgs` struct to an `ast` node and remove step 3, moving that step to later in the compilation process.

It also fixes a few diagnostics bugs.

This also [significantly reduces](https://gist.github.com/m-ou-se/b67b2d54172c4837a5ab1b26fa3e5284) the amount of generated code for cases with arguments in non-default order without formatting options, like `"{1} {0}"` or `"{a} {}"`, etc.

### CODE REVIEW

**Comment by m-ou-se** [comment]
@Alexendoo I saw you've sent a few PRs touching this code recently. In case you were planning to send more of those, be aware I'm refactoring the entire thing. ^^

**Comment by Alexendoo** [comment]
Thanks for the heads up! The main one I was planning to do #101000 so that is handy 😄

**Comment by m-ou-se** [comment]
r? rust-lang/compiler

**Comment by m-ou-se** [comment]
@estebank The bot selected you, but feel free to reassign if you think there's another reviewer who haqs more context here. It's a big PR.

I'd also be happy to go through this PR in a live chat or video call, if that's useful. :)

File: src/test/ui/fmt/ifmt-bad-arg.stderr [hunk 1]
```diff
@@ -5,10 +5,10 @@ LL |     format!("{}");
    |              ^^
 
 error: invalid reference to positional argument 1 (there is 1 argument)
```
Comment (estebank) [inline src/test/ui/fmt/ifmt-bad-arg.stderr:7]: Does this need a `bless`ing or there are some cases where the message hasn't changed?

File: src/test/ui/fmt/format-args-capture-issue-93378.stderr [hunk 2]
```diff
@@ -1,19 +1,14 @@
-error: invalid reference to positional arguments 1 and 2 (there is 1 argument)
-  --> $DIR/format-args-capture-issue-93378.rs:5:26
+error: 3 positional arguments in format string, but there is 1 argument
```
Comment (estebank) [inline src/test/ui/fmt/format-args-capture-issue-93378.stderr:1]: Should we refer to the positional/named arguments as "formatting arguments"?

```suggestion
error: 3 positional arguments in format string, but there is 1 formatting named argument
```

File: src/test/ui/fmt/format-args-capture-issue-93378.stderr [hunk 3]
```diff
@@ -1,19 +1,14 @@
...
+  --> $DIR/format-args-capture-issue-93378.rs:5:23
    |
 LL |     println!("{a} {b} {} {} {c} {}", c = "c");
-   |                          ^^     ^^
-   |
-   = note: positional arguments are zero-based
+   |                       ^^ ^^     ^^       ---
```
Comment (estebank) [inline src/test/ui/fmt/format-args-capture-issue-93378.stderr:5]:
```suggestion
   |                       ^^ ^^     ^^       --- only 1 named formatting argument available
   |                       |
   |                       3 positional arguments in the format string
```

File: src/test/ui-fulldeps/session-diagnostic/diagnostic-derive.stderr [hunk 4]
```diff
@@ -212,21 +212,19 @@ LL |     #[suggestion(typeck::suggestion, code = "{name}")]
...
 
 error: invalid format string: expected `'}'` but string was terminated
-  --> $DIR/diagnostic-derive.rs:171:16
+  --> $DIR/diagnostic-derive.rs:171:10
    |
 LL | #[derive(SessionDiagnostic)]
-   |           -    ^ expected `'}'` in format string
-   |           |
-   |           because of this opening brace
+   |          ^^^^^^^^^^^^^^^^^ expected `'}'` in format string
```
Comment (estebank) [inline src/test/ui-fulldeps/session-diagnostic/diagnostic-derive.stderr:218]: Lovely drive-by fix 😅

File: compiler/rustc_builtin_macros/src/format/ast.rs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    /// `{}` or `{.*}`
+    Implicit,
+    /// `{1}` or `{:1$}` or `{:.1$}`
+    Number,
+    /// `{a}` or `{:a$}` or `{:.a$}`
+    Named,
+}
+
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
+pub enum FormatTrait {
```
Comment (estebank) [inline compiler/rustc_builtin_macros/src/format/ast.rs:108]: Should we add a `Custom(String)` variant that we error out on, but that might be allowed in the future?

File: compiler/rustc_builtin_macros/src/format/ast.rs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    /// The width. E.g. `{:5}` or `{:width$}`.
+    pub width: Option<FormatCount>,
+    /// The precision. E.g. `{:.5}` or `{:.precision$}`.
+    pub precision: Option<FormatCount>,
+    /// The alignment. E.g. `{:>}` or `{:<}` or `{:^}`.
+    pub alignment: Option<FormatAlignment>,
+    /// The fill character. E.g. the `.` in `{:.>10}`.
+    pub fill: Option<char>,
+    /// The `+`, `-`, `0`, `#`, `x?` and `X?` flags.
+    pub flags: u32,
```
Comment (estebank) [inline compiler/rustc_builtin_macros/src/format/ast.rs:140]: This is regrettably non-type-safe. I'd prefer we had a newtype that wraps over `u32`, exposes the valid "variants" and implements `|`. (I know I'm asking more than what was already there.)

File: compiler/rustc_builtin_macros/src/format.rs [hunk 7]
```diff
@@ -1,271 +1,42 @@
...
+//  1. First, `parse_args` will parse the `(literal, arg, arg, name=arg, name=arg)` syntax,
+//     but doesn't parse the template (the literal) itself.
+//  2. Second, `make_format_args` will parse the template, the format options, resolve argument references,
+//     produce diagnostics, and turn the whole thing into a `FormatArgs` structure.
+//  3. Finally, `expand_parsed_format_args` will turn that `FormatArgs` structure
+//     into the expression that the macro expands to.
 
-    /// Starting offset of count argument slots.
-    count_args_index_offset: usize,
+// See format/ast.rs forthe FormatArgs structure.
```
Comment (estebank) [inline compiler/rustc_builtin_macros/src/format.rs:29]:
```suggestion
// See format/ast.rs for the FormatArgs structure.
```

Should we also mention the glossary you wrote out?

File: compiler/rustc_builtin_macros/src/format.rs [hunk 8]
```diff
@@ -1,271 +1,42 @@
...
-    name: Option<Ident>,
+// Only used in parse_args and report_invalid_references,
+// to indicate how a referred argument was used.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum PositionUsedAs {
+    Placeholder(Option<Span>),
+    Precision,
+    Width,
 }
+use PositionUsedAs::*;
```
Comment (estebank) [inline compiler/rustc_builtin_macros/src/format.rs:39]: I think this `use` after the definition might go against style. This could also be at the top of the file and work correctly.

File: compiler/rustc_builtin_macros/src/format.rs [hunk 9]
```diff
@@ -347,21 +117,19 @@ fn parse_args<'a>(
                 p.bump();
                 p.expect(&token::Eq)?;
                 let e = p.parse_expr()?;
-                if let Some(&prev) = names.get(&ident.name) {
-                    ecx.struct_span_err(e.span, &format!("duplicate argument named `{}`", ident))
-                        .span_label(args[prev].expr.span, "previously here")
-                        .span_label(e.span, "duplicate argument")
-                        .emit();
+                if let Some(prev) =
+                    args.iter().rev().map_while(|a| a.1.ident()).find(|n| n.name == ident.name)
```
Comment (estebank) [inline compiler/rustc_builtin_macros/src/format.rs:121]: Shouldn't we instead "waste" some memory keeping a `HashSet<Symbol>` or `HashSet<Ident>` and checking that here? This is O(n^2) *in the happy path* for a case that I'm pretty sure isn't tracked in the perf suite.

File: compiler/rustc_builtin_macros/src/format.rs [hunk 10]
```diff
@@ -371,851 +139,31 @@ fn parse_args<'a>(
                         "positional arguments cannot follow named arguments",
                     );
                     err.span_label(e.span, "positional arguments must be before named arguments");
```
Comment (estebank) [inline compiler/rustc_builtin_macros/src/format.rs:141]: What's the output for `format!("{x} {} {} {}", x = 1, 2, 3);`? I believe we'll have at least 2 errors for the non named args, right? [That's the current behavior](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c7e2cfc3df33a999f11a009a40341373). Should we instead try and accumulate all of these and emit a single error like

```
error: positional arguments cannot follow named arguments
 --> src/main.rs:2:36
  |
2 |     format!("{x} {} {} {}", x = 1, 2, 3);
  |                                 -  ^  ^ positional arguments must be before named arguments
  |                                 |
  |                                 named argument
```

File: compiler/rustc_builtin_macros/src/format.rs [hunk 11]
```diff
@@ -371,851 +139,31 @@ fn parse_args<'a>(
...
-    // NOTE: this verbose way of initializing `Vec<Vec<ArgumentType>>` is because
-    // `ArgumentType` does not derive `Clone`.
-    let arg_types: Vec<_> = (0..args.len()).map(|_| Vec::new()).collect();
-    let arg_unique_types: Vec<_> = (0..args.len()).map(|_| Vec::new()).collect();
-
-    let mut macsp = ecx.call_site();
-    macsp = ecx.with_def_site_ctxt(macsp);
+) -> Result<FormatArgs, ()> {
+    let start_of_named_args =
+        args.iter().position(|arg| arg.1.ident().is_some()).unwrap_or(args.len());
```
Comment (estebank) [inline compiler/rustc_builtin_macros/src/format.rs:163]: This is also something that you would have found out by now to be able to report the error of a positional arg coming after a named arg, right? That's another O(n) case that we can eliminate.

File: compiler/rustc_builtin_macros/src/format.rs [hunk 12]
```diff
@@ -1287,266 +237,671 @@ pub fn expand_preparsed_format_args(
             e.note(&note);
         }
         if let Some((label, span)) = err.secondary_label {
-            if efmt_kind_is_lit {
+            if is_literal {
```
Comment (estebank) [inline compiler/rustc_builtin_macros/src/format.rs:240]: This can now be a let chain

File: compiler/rustc_builtin_macros/src/format.rs [hunk 13]
```diff
@@ -1287,266 +237,671 @@ pub fn expand_preparsed_format_args(
...
-        fmtsp: fmt_span,
-        invalid_refs: Vec::new(),
-        arg_spans,
-        arg_with_formatting: Vec::new(),
-        is_literal: parser.is_literal,
-        unused_names_lint: PositionalNamedArgsLint { positional_named_args: vec![] },
+    enum ArgRef<'a> {
+        Index(usize),
+        Name(&'a str, Option<Span>),
+    }
```
Comment (estebank) [inline compiler/rustc_builtin_macros/src/format.rs:275]: Should we declare top level instead of inside this function?

File: compiler/rustc_builtin_macros/src/format.rs [hunk 14]
```diff
@@ -1287,266 +237,671 @@ pub fn expand_preparsed_format_args(
...
+    use ArgRef::*;
+
+    let num_explicit_args = args.len();
```
Comment (estebank) [inline compiler/rustc_builtin_macros/src/format.rs:278]: Should this be `num_args`? Because it implicitly includes the named args as well.

File: compiler/rustc_builtin_macros/src/format.rs [hunk 15]
```diff
@@ -1287,266 +237,671 @@ pub fn expand_preparsed_format_args(
...
+                            // Doesn't exist as an explicit argument.
+                            invalid_refs.push((index, span, used_as, kind));
+                            Err(index)
+                        }
+                    }
+                } else {
+                    invalid_refs.push((index, span, used_as, kind));
+                    Err(index)
+                }
+            }
```
Comment (estebank) [inline compiler/rustc_builtin_macros/src/format.rs:311]: I'm looking at how we could remove some levels of indentation. What about this?

```
            Index(index) => {
                match args.get(index) {
                    Some((_, FormatArgKind::Normal) => {
                        used[index] = true;
                        Ok(index)
                    }
                    Some((_, FormatArgKind::Named(_)) => {
                        used[index] = true;
                        numeric_refences_to_named_arg.push((index, span, used_as));
                        Ok(index)
                    }
                    FormatArgKind::Captured(_)
                        // Doesn't exist as an explicit argument.
                    | None => {
                        invalid_refs.push((index, span, used_as, kind));
                        Err(index)
                    }
                }
            }
```

File: compiler/rustc_builtin_macros/src/format.rs [hunk 16]
```diff
@@ -1287,266 +237,671 @@ pub fn expand_preparsed_format_args(
...
-    let pieces = unverified_pieces
-        .into_iter()
-        .map(|mut piece| {
-            cx.verify_piece(&piece);
-            cx.resolve_name_inplace(&mut piece);
-            piece
-        })
-        .collect::<Vec<_>>();
+    let mut template = Vec::new();
+    let mut unfinished_literal = String::new();
```
Comment (estebank) [inline compiler/rustc_builtin_macros/src/format.rs:346]: Would using `with_capacity` be worth it here? 🤔

File: src/test/ui/fmt/ifmt-bad-arg.stderr [hunk 17]
```diff
@@ -307,10 +291,10 @@ error[E0308]: mismatched types
...
    |
 LL |     println!("{} {:.*} {}", 1, 3.2, 4);
-   |     ---------------------------^^^----
-   |     |                          |
-   |     |                          expected `usize`, found floating-point number
-   |     arguments to this function are incorrect
+   |                                ^^^
+   |                                |
+   |                                expected `usize`, found floating-point number
+   |                                arguments to this function are incorrect
```
[the remark this answers is not in the document]
Comment (m-ou-se) [reply src/test/ui/fmt/ifmt-bad-arg.stderr:297]: That's just from the `mismatched types` error:

```
error[E0308]: mismatched types
 --> src/main.rs:2:23
  |
2 |     fn a(_: i32) {} a("asdf");
  |                     - ^^^^^^ expected `i32`, found `&str`
  |                     |
  |                     arguments to this function are incorrect
  |
```

It tries to point at the `ArgumentV1::from_usize(..)` call and the argument to that call, which both have the same span.

File: src/test/ui/fmt/ifmt-bad-arg.stderr (hunk 1 above)
Comment (m-ou-se) [reply src/test/ui/fmt/ifmt-bad-arg.stderr:7]: This is about `format!("{1}", 1)`. It complains about a reference to `1` when there's only one argument. That remains unchanged.

File: src/test/ui/fmt/format-args-capture-issue-93378.stderr (hunk 3 above)
Comment (m-ou-se) [reply src/test/ui/fmt/format-args-capture-issue-93378.stderr:5]: That seems like something for a separate PR. This refactor is meant as a starting point to make it much easier to make changes like that afterwards, since now the part that does the diagnostics is completely independent of the representation of the expanded `fmt::Arguments`.

For now, I tried to keep the diagnostics mostly unchanged, except for a few bug fixes, and a few cases were it now picks the rignt alternative because the right information is now available. (E.g. when to pick the simple "{N} positional arguments in format string, but .." diagnostic instead of the more general "invalid refernce to positional argument {X}" one.)

File: compiler/rustc_builtin_macros/src/format/ast.rs (hunk 6 above)
Comment (m-ou-se) [reply compiler/rustc_builtin_macros/src/format/ast.rs:140]: Yeah, that's something I still want to change. But that `u32` comes from the separate `rustc_parse_format` crate, which isn't part of this refactor. We first need to decouple std's `fmt::Arguments` and rustc's `rustc_parse_format` a bit more.

File: compiler/rustc_builtin_macros/src/format.rs (hunk 9 above)
Comment (m-ou-se) [reply compiler/rustc_builtin_macros/src/format.rs:121]: Maybe. A linear search through a short Vec is extremely fast though. I'm not sure how often we'd have a `format_args!()` so big that the `O(..)` behavior becomes relevant.

File: compiler/rustc_builtin_macros/src/format.rs (hunk 10 above)
Comment (m-ou-se) [reply compiler/rustc_builtin_macros/src/format.rs:141]: The current behavior is unchanged in this PR. (Except the span is now for the entire `x = 1` instead of just `1`.) We can improve it in a separate PR.

File: compiler/rustc_builtin_macros/src/format.rs (hunk 14 above)
Comment (m-ou-se) [reply compiler/rustc_builtin_macros/src/format.rs:278]: It only includes the explicitly given arguments, not the implicit/captured ones.

For `format!("{} {a} {b}", 1, a = 2)`, it is 2.

(The `Captured` argument for `b` is added by `lookup_arg` below.)

File: compiler/rustc_builtin_macros/src/format.rs (hunk 16 above)
Comment (m-ou-se) [reply compiler/rustc_builtin_macros/src/format.rs:346]: The first `.push()` to an empty `Vec` and the first `.push_str()` to an empty `String` will pick a reasonable capacity for that first allocation.

File: compiler/rustc_builtin_macros/src/format.rs (hunk 8 above)
Comment (m-ou-se) [reply compiler/rustc_builtin_macros/src/format.rs:39]: I think it'd look more out of place there. I put it here for the same reason I put the `mod`s together with their `use`, a few lines above:

```rust
mod ast;
use ast::*;

mod expand;
use expand::expand_parsed_format_args;
```

File: compiler/rustc_builtin_macros/src/format.rs [hunk 18]
```diff
@@ -1287,266 +237,671 @@ pub fn expand_preparsed_format_args(
...
+            }
         }
     }
 
-    if !cx.literal.is_empty() {
-        let s = cx.build_literal_string();
-        cx.str_pieces.push(s);
+    if !unfinished_literal.is_empty() {
+        template.push(FormatArgsPiece::Literal(Symbol::intern(&unfinished_literal)));
+        drop(unfinished_literal);
```
Comment (estebank) [inline compiler/rustc_builtin_macros/src/format.rs:469]: Why's this `drop` needed?

Comment (m-ou-se) [reply compiler/rustc_builtin_macros/src/format.rs:462]: That way that variable stops existing. It makes it clear to the reader that the `unfinished_literal` variable is not used afterwards, even though the scope in which it was defined isn't finished yet. (The compiler would error if we try to use `unfinished_literal` afterwards.)

I could've used a new `{}` scope, but that'd have increased indentation again.

File: compiler/rustc_builtin_macros/src/format.rs [hunk 19]
```diff
@@ -1287,266 +237,671 @@ pub fn expand_preparsed_format_args(
...
+                        // disabled (see RFC #2795)
+                        ecx.struct_span_err(span, &format!("there is no argument named `{name}`"))
+                            .note(format!("did you intend to capture a variable `{name}` from the surrounding scope?"))
+                            .note("to avoid ambiguity, `format_args!` cannot capture variables when the format string is expanded from a macro")
+                            .emit();
+                        DummyResult::raw_expr(span, true)
+                    };
+                    args.push((arg, FormatArgKind::Captured(ident)));
+                    Ok(args.len() - 1)
+                }
```
[the remark this answers is not in the document]
Comment (m-ou-se) [reply compiler/rustc_builtin_macros/src/format.rs]: Added a few comments.

File: src/test/ui/fmt/format-args-capture-issue-93378.stderr (hunk 2 above)
Comment (estebank) [reply src/test/ui/fmt/format-args-capture-issue-93378.stderr:1]: Let's deal with the wording on a follow up PR then.

File: src/test/ui/fmt/format-args-capture-issue-93378.stderr (hunk 3 above)
Comment (estebank) [reply src/test/ui/fmt/format-args-capture-issue-93378.stderr:5]: Fair enough. I started looking at the diagnostics first to get a sense for what might have changed, which has the annoying side effect of me noticing all of these things I'd like to clean up regardless 😅

File: compiler/rustc_builtin_macros/src/format.rs (hunk 9 above)
Comment (estebank) [reply compiler/rustc_builtin_macros/src/format.rs:121]: It's not so much about how bad it is in the general case, it is about resiliency in the extremes. I could see for example a templating engine ending up being a big `format` call with a ton of arguments.

File: compiler/rustc_builtin_macros/src/format.rs (hunk 11 above)
Comment (estebank) [reply compiler/rustc_builtin_macros/src/format.rs:163]: I agree that this handles reasonable cases perfectly, but there is a lot of unreasonable out there 😅

I figured that you were trying to minimize coupling, and feels reasonable. We might land as is and revisit later, particularly in face of the HIR change you'd proposed?

File: compiler/rustc_builtin_macros/src/format.rs (hunk 18 above)
Comment (estebank) [reply compiler/rustc_builtin_macros/src/format.rs:465]: Why not drop it unconditionally then?

File: compiler/rustc_builtin_macros/src/format/ast.rs (hunk 5 above)
Comment (m-ou-se) [reply compiler/rustc_builtin_macros/src/format/ast.rs:108]: What errors do you mean? Like, when someone does `{:z}` or something, to propagate that `"z"`?

File: compiler/rustc_builtin_macros/src/format.rs (hunk 9 above)
Comment (m-ou-se) [reply compiler/rustc_builtin_macros/src/format.rs:121]: Ah right, that makes sense.

I'll change the `Vec<(P<Expr>, FormatArgKind)>` to a new struct that represents the arguments in a way that allows quick lookup through a hashmap. (That also resolves your comment about `start_of_named_args` below.)

(Not sure what to name that struct though. :thinking: `Arguments`? `FormatArguments`? `ArgumentsToFormatArgs`? `FormatArgsArguments`? `FormatArgsArgs`? Confusing.)

File: compiler/rustc_builtin_macros/src/format/ast.rs (hunk 5 above)
Comment (estebank) [reply compiler/rustc_builtin_macros/src/format/ast.rs:108]: Exactly, but let's not deal with that now.

File: compiler/rustc_builtin_macros/src/format.rs (hunk 9 above)
Comment (m-ou-se) [reply compiler/rustc_builtin_macros/src/format.rs]: Updated.

**Comment by m-ou-se** [comment]
@estebank This is ready for review again. :)

**Comment by estebank** [comment]
Apologies for the delay, this is in my queue to re-review.

File: compiler/rustc_builtin_macros/src/format.rs [hunk 20]
```diff
@@ -1286,267 +232,648 @@ pub fn expand_preparsed_format_args(
...
+        "the only appropriate formatting traits are:\n\
+                                - ``, which uses the `Display` trait\n\
+                                - `?`, which uses the `Debug` trait\n\
+                                - `e`, which uses the `LowerExp` trait\n\
+                                - `E`, which uses the `UpperExp` trait\n\
+                                - `o`, which uses the `Octal` trait\n\
+                                - `p`, which uses the `Pointer` trait\n\
+                                - `b`, which uses the `Binary` trait\n\
+                                - `x`, which uses the `LowerHex` trait\n\
+                                - `X`, which uses the `UpperHex` trait",
```
Comment (estebank) [inline compiler/rustc_builtin_macros/src/format.rs:530]: This feels like something we should build from a source of truth, to avoid potentially forgetting to update this in the future.

**Review by estebank — APPROVED** [review]
r=me after rebasing

**Comment by m-ou-se** [comment]
Weird, the relevant logs seem to be empty.

@bors retry

**Comment by nnethercote** [comment]
`keccak` is noisy at the moment. The other changes are likely real, but the wins and losses balance out.

@rustbot label: +perf-regression-triaged
